lib/ide/src/double_representation/definition.rs

```diff
@@ -0,0 +1,272 @@
+//! Code for definition discovery in the blocks, finding definition by name and related utilities.
+
+use crate::prelude::*;
+
+use ast::Ast;
+use ast::HasRepr;
+use ast::Shape;
+use ast::known;
+use ast::prefix;
+use ast::opr;
+
+/// Describes the kind of code block (scope) to which definition can belong.
+#[derive(Clone,Copy,Debug,PartialEq)]
+pub enum ScopeKind {
+    /// Module scope is a file's top-level block.
+    Root,
+    /// Any other block, e.g. introduced as body of some definition binding.
+    NonRoot,
+}
+
+
+
+// ==================
+// === Identifier ===
+// ==================
+
+/// Checks if given Ast node can be used to represent identifier being part of definition name.
+pub fn is_identifier(ast:&Ast) -> bool {
+    match ast.shape() {
+        Shape::Var          {..} => true,
+        Shape::Cons         {..} => true,
+        Shape::SectionSides {..} => true,
+        Shape::Opr          {..} => true,
+        _                        => false,
+    }
+}
+
+/// Retrieves the identifier's name, if the Ast node is an identifier. Otherwise, returns None.
+pub fn identifier_name(ast:&Ast) -> Option<String> {
+    is_identifier(ast).then_with(|| ast.repr())
+}
+
+
+
+// ======================
+// === DefinitionName ===
+// ======================
+
+/// Structure representing definition name. If this is an extension method, extended type is
+/// also included.
+#[derive(Clone,Debug,PartialEq)]
+pub struct DefinitionName {
+    /// Used when definition is an extension method. Then it stores the segments
+    /// of the extended target type path.
+    pub extended_target : Vec<String>,
+    /// Name of the function itself.
+    pub name : String,
+}
+
+impl DefinitionName {
+    /// Creates a new name consisting of a single identifier, without any extension target.
+    pub fn new_plain(name:impl Str) -> DefinitionName {
+        DefinitionName {name:name.into(), extended_target:default()}
+    }
+
+    /// Tries describing given Ast piece as a definition name. Typically, passed Ast
+    /// should be the binding's left-hand side.
+    ///
+    /// Returns `None` if is not name-like entity.
+    pub fn from_ast(ast:&Ast) -> Option<DefinitionName> {
+        let accessor_chain = opr::Chain::try_new_of(ast,opr::predefined::ACCESS);
+        let (extended_target,name) = match accessor_chain {
+            Some(accessor_chain) => {
+                let mut args = vec![identifier_name(&accessor_chain.target?)?];
+                for arg in accessor_chain.args.iter() {
+                    let arg_ast = arg.1.as_ref()?;
```

Comment:
And now I have to check what the 1 means. Can we create structure with named field instead of tuple there?